### PR TITLE
WIP: Release candidate 1.0.0

### PR DESCRIPTION
**Pull request by fedden**
Not ready for review yet, but this PR will be the one that brings us to version 1.0.0, and enables our first release.

To do:
- [X] Change module name
- [X] Unify CFR and other core algos
- [X] Add multithreaded CFR code to module
- [X] Remove defunct scripts
- [x] Create root CLI script
- [x] Document root CLI script
- [ ] Create base state class
- [ ] Add examples dir with examples of key classes etc
- [ ] Update default branch to master
- [x] Add HISTORY.md
- [x] Update roadmap in README.md
- [x] Add pypi package
- [x] Setup github actions CD to push package to pypi whenever we make a new release
- [ ] Setup github actions CD to build documentation whenever we make a new release
- [x] Write test for multithreaded CFR code
- [x] Add singlethreaded CFR code to module
- [x] Write test for multithreaded CFR code
- [ ] Add RT code
- [X] Add terminal app to play against bot
- [ ] Test coverage >= 80%
- [ ] Clean up code base (mostly OG poker_engine, other stuff)
- [ ] Add quantised dict class
- [x] Add auto documentation for docstrings
- [x] Add documentation on how to build the documentation
- [ ] Add tutorial on how to call the various scripts, in ReadTheDocs

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -4,19 +4,19 @@
...
-    description="Open source implementation of the pluribus poker AI ployer.",
+    description="Open source implementation of a CFR based poker AI player.",
     long_description=long_description,
     long_description_content_type="text/markdown",
-    url="https://github.com/fedden/pluribus",
+    url="https://github.com/fedden/poker_ai",
     packages=setuptools.find_packages(),
     classifiers=[
         "Programming Language :: Python :: 3",
         "License :: OSI Approved :: MIT License",
```
Comment (big-c-note) [inline setup.py:18]: FYI, just noticing this. @big-c-note

File: poker_ai/ai/multiprocess/server.py [hunk 2]
```diff
@@ -68,7 +68,7 @@ def __init__(
         self._worker_status: Dict[str, str] = dict()
         self._agent: Agent = Agent(agent_path)
         self._locks: Dict[str, mp.synchronize.Lock] = dict(
-            regret=mp.Lock(), strategy=mp.Lock()
+            regret=mp.Lock(), strategy=mp.Lock(), pre_flop_strategy=mp.Lock()
```
Comment (big-c-note) [inline poker_ai/ai/multiprocess/server.py:71]: Please check me on this, seems to work as expected.

File: poker_ai/ai/multiprocess/server.py [hunk 3]
```diff
@@ -86,7 +86,7 @@ def search(self):
         progress_bar = progress_bar_manager.counter(
             total=self._n_iterations, desc="Optimisation iterations", unit="iter"
         )
-        for t in range(self._start_timestep, self._n_iterations):
+        for t in range(self._start_timestep, self._n_iterations + 1):
```
Comment (big-c-note) [inline poker_ai/ai/multiprocess/server.py:89]: Updating for dump iteration. If n_iterations is divisible by the dump_int, then we dump the final iteration.

File: poker_ai/ai/ai.py [hunk 4]
```diff
@@ -68,8 +68,8 @@ def update_strategy(
         sigma = calculate_strategy(this_info_sets_regret)
         log.debug(f"Calculated Strategy for {state.info_set}: {sigma}")
         # choose an action based of sigma
-        available_actions: List[str] = sigma["actions"]
-        action_probabilities: np.ndarray = sigma["probabilities"]
+        available_actions: List[str] = list(sigma.keys())
```
Comment (big-c-note) [inline poker_ai/ai/ai.py:71]: Fix update_strategy bug

File: poker_ai/ai/ai.py [hunk 5]
```diff
@@ -303,6 +308,11 @@ def serialise(
     offline_agent["regret"] = copy.deepcopy(agent.regret)
     if locks:
         locks["regret"].release()
+    if locks:
```
Comment (big-c-note) [inline poker_ai/ai/ai.py:311]: Dump pre flop strategy

File: poker_ai/clustering/runner.py [hunk 6]
```diff
@@ -290,7 +289,7 @@ def simulate_get_turn_ehs_distributions(
         # sample river cards and run a simulation
         for _ in range(num_simulations):
             river_card = np.random.choice(available_cards, 1, replace=False)
-            board = the_board + river_card
+            board = np.append(the_board, river_card)
```
Comment (big-c-note) [inline poker_ai/clustering/runner.py:292]: These are numpy arrays.

File: poker_ai/clustering/runner.py [hunk 7]
```diff
@@ -328,7 +327,7 @@ def get_available_cards(
         """Get all cards that are available."""
         # Turn into set for O(1) lookup speed.
         unavailable_cards = set(unavailable_cards.tolist())
-        return np.ndarray([c for c in cards if c not in unavailable_cards])
+        return np.array([c for c in cards if c not in unavailable_cards])
```
Comment (big-c-note) [inline poker_ai/clustering/runner.py:330]: This isn't multi dimensional

File: poker_ai/clustering/runner.py [hunk 8]
```diff
@@ -357,7 +356,7 @@ def process_flop_potential_aware_distributions(
         available_cards: np.ndarray = self.get_available_cards(
             cards=self._cards, unavailable_cards=public
         )
-        potential_aware_distribution_flop = np.zeros_like(self._turn_centroids)
+        potential_aware_distribution_flop = [0] * len(self._turn_centroids)
```
Comment (big-c-note) [inline poker_ai/clustering/runner.py:359]: Returning a numpy array from this method when this object was also array created a 3d matrix that the sklearn clustering couldn't support.

File: poker_ai/clustering/runner.py [hunk 9]
```diff
@@ -161,9 +161,12 @@ def __init__(self, save_dir: str = ""):
         """"""
         super().__init__()
         self.card_info_lut_path: Path = Path(save_dir) / "card_info_lut.joblib"
+        self.centroid_path: Path = Path(save_dir) / "centroids.joblib"
```
Comment (big-c-note) [inline poker_ai/clustering/runner.py:182]: Need to dump the centroids as they are used in clustering the next streets.

File: poker_ai/clustering/runner.py [hunk 10]
```diff
@@ -400,6 +406,14 @@ def cluster(num_clusters: int, X: np.array):
         centroids = km.cluster_centers_
         return centroids, y_km
 
+    @staticmethod
```
Comment (big-c-note) [inline poker_ai/clustering/runner.py:482]: Need to turn the data into the lookup table before dumping.

File: poker_ai/ai/multiprocess/server.py (hunk 2 above)
Comment (fedden) [reply poker_ai/ai/multiprocess/server.py:71]: all good

**Pull request merged**